File: agent/droplet_agent/pinning.py
"""Trusting the hub's self-signed LAN certificate by its fingerprint.

On the LAN the hub serves HTTPS with a long-lived self-signed certificate
(docs/local-first.md §2). The agent accepts it if and only if the SHA-256 of
its DER encoding equals the fingerprint it pinned when it paired. Hostname and
CA checks are off for these connections only; the tailnet URL keeps ordinary,
fully verified TLS.

The check runs inside `wrap_socket`, right after the TLS handshake and before
a single byte of HTTP is written. So a machine that isn't the hub never sees
the device token: not in a request, not in the WebSocket upgrade.
"""

from __future__ import annotations

import hashlib
import hmac
import re
import ssl

FINGERPRINT = re.compile(r"^[0-9a-f]{64}$")


class HubError(Exception):
    """Talking to the hub failed. `pair` is set when the hub says this device isn't let in."""

    def __init__(self, message: str, *, pair: bool = False):
        super().__init__(message)
        self.pair = pair


class PinMismatch(HubError):
    """The machine at a LAN address presented a certificate other than the pinned one."""

    def __init__(self, where: str, expected: str, got: str | None):
        self.where = where
        self.expected = expected
        self.got = got
        super().__init__(
            f"the hub's identity changed: {where} presented a certificate with fingerprint "
            f"{got or 'none'}, not the pinned {expected}. Either that isn't your hub, or its "
            "certificate was regenerated. Nothing was sent to it. If you trust it, re-pair "
            "with: droplet-agent setup")


def fingerprint(der: bytes) -> str:
    """SHA-256 of a DER certificate, lowercase hex: the form /api/hub/info and mDNS use."""
    return hashlib.sha256(der).hexdigest()


def normalize_fingerprint(fp) -> str | None:
    """A valid 64-hex fingerprint, lowercased, or None. Colons (openssl's format) are dropped."""
    if not isinstance(fp, str):
        return None
    fp = fp.replace(":", "").strip().lower()
    return fp if FINGERPRINT.match(fp) else None


def check_peer(sock: ssl.SSLSocket, pin: str, where: str = "the hub"):
    """Raise PinMismatch unless the connected peer's certificate matches `pin`."""
    der = sock.getpeercert(binary_form=True)
    got = fingerprint(der) if der else None
    if got is None or not hmac.compare_digest(got, pin):
        raise PinMismatch(where, pin, got)


class PinnedContext(ssl.SSLContext):
    """A client context that trusts exactly one certificate, by fingerprint.

    urllib (http.client) and websockets both call `wrap_socket` on an already
    connected socket, so the handshake completes in there and the pin is
    checked before either of them sends anything.
    """

    pin: str

    def wrap_socket(self, sock, *args, **kwargs):
        kwargs["do_handshake_on_connect"] = True
        tls = super().wrap_socket(sock, *args, **kwargs)
        try:
            try:
                where = "%s:%s" % tls.getpeername()[:2]
            except OSError:
                where = "the hub"
            check_peer(tls, self.pin, where)
        except BaseException:
            tls.close()
            raise
        return tls


def pinned_context(pin: str) -> PinnedContext:
    good = normalize_fingerprint(pin)
    if good is None:
        raise ValueError(f"not a SHA-256 fingerprint: {pin!r}")
    ctx = PinnedContext(ssl.PROTOCOL_TLS_CLIENT)
    ctx.check_hostname = False          # the certificate is pinned instead
    ctx.verify_mode = ssl.CERT_NONE     # a self-signed certificate has no CA to check
    ctx.minimum_version = ssl.TLSVersion.TLSv1_2
    ctx.pin = good
    return ctx

File: android/app/src/main/java/dev/droplet/app/Pinning.kt
package dev.droplet.app

import android.net.http.SslCertificate
import android.os.Build
import android.os.Bundle
import okhttp3.OkHttpClient
import java.io.ByteArrayInputStream
import java.net.InetSocketAddress
import java.security.MessageDigest
import java.security.cert.Certificate
import java.security.cert.CertificateException
import java.security.cert.CertificateFactory
import java.security.cert.X509Certificate
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit
import javax.net.ssl.HostnameVerifier
import javax.net.ssl.SSLContext
import javax.net.ssl.SSLSession
import javax.net.ssl.SSLSocket
import javax.net.ssl.X509TrustManager

/**
 * The hub's LAN certificate is self-signed, so on the LAN the app trusts it
 * by its fingerprint and nothing else (docs/local-first.md §2): a connection
 * is accepted if and only if the server's leaf certificate hashes (SHA-256,
 * DER) to the pinned value. There are no CA or hostname checks, because the
 * hub is reached by an IP that DHCP hands out.
 *
 * The tailnet keeps OkHttp's normal, fully verified client.
 */
object Pinning {
    /** A server presented a certificate other than the pinned one. */
    class Mismatch(val expected: String, val seen: String) :
        CertificateException("The hub's certificate doesn't match the one this phone paired with")

    /** Thrown on purpose by [capture] once the certificate is in hand. */
    private class Captured(val fingerprint: String) : CertificateException("captured")

    fun sha256(der: ByteArray): String =
        MessageDigest.getInstance("SHA-256").digest(der).joinToString("") { "%02x".format(it) }

    fun fingerprint(cert: Certificate): String = sha256(cert.encoded)

    /** A well-formed pin: 64 lowercase hex characters. */
    fun isFingerprint(s: String?): Boolean = s != null && s.length == 64 && s.all { it in '0'..'9' || it in 'a'..'f' }

    /** Accepts exactly one leaf certificate. Everything else is refused. */
    class PinnedTrustManager(private val pin: String) : X509TrustManager {
        init {
            require(isFingerprint(pin)) { "not a SHA-256 fingerprint" }
        }

        override fun checkServerTrusted(chain: Array<out X509Certificate>?, authType: String?) {
            val leaf = chain?.firstOrNull() ?: throw CertificateException("The server sent no certificate")
            val seen = fingerprint(leaf)
            // constant time, although the pin is no secret
            if (!MessageDigest.isEqual(seen.toByteArray(), pin.toByteArray())) throw Mismatch(pin, seen)
        }

        override fun checkClientTrusted(chain: Array<out X509Certificate>?, authType: String?) {
            throw CertificateException("This trust manager only checks the hub")
        }

        override fun getAcceptedIssuers(): Array<X509Certificate> = emptyArray()
    }

    /**
     * Stands in for the hostname check: the peer's leaf must be the pinned
     * certificate. The trust manager has already checked that; this is the
     * second lock, so neither can be loosened alone.
     */
    class PinnedHostnameVerifier(private val pin: String) : HostnameVerifier {
        override fun verify(hostname: String?, session: SSLSession?): Boolean {
            val leaf = try {
                session?.peerCertificates?.firstOrNull()
            } catch (e: Exception) {
                null
            } ?: return false
            return MessageDigest.isEqual(fingerprint(leaf).toByteArray(), pin.toByteArray())
        }
    }

    private val clients = ConcurrentHashMap<String, OkHttpClient>()

    /** A client that talks only to the server holding [pin]'s certificate, sharing [base]'s pool and settings. */
    fun client(pin: String, base: OkHttpClient = Hub.client): OkHttpClient {
        if (base === Hub.client) clients[pin]?.let { return it }
        val tm = PinnedTrustManager(pin)
        val ctx = SSLContext.getInstance("TLS").apply { init(null, arrayOf(tm), null) }
        val built = base.newBuilder()
            .sslSocketFactory(ctx.socketFactory, tm)
            .hostnameVerifier(PinnedHostnameVerifier(pin))
            // a LAN hub answers quickly or not at all (the phone left the Wi-Fi)
            .connectTimeout(LAN_CONNECT_S, TimeUnit.SECONDS)
            .build()
        if (base === Hub.client) clients[pin] = built
        return built
    }

    /** The [Mismatch] behind a failed connection, if that's why it failed. */
    fun mismatch(t: Throwable?): Mismatch? {
        var e = t
        var depth = 0
        while (e != null && depth++ < 10) {
            if (e is Mismatch) return e
            e.suppressed.firstNotNullOfOrNull { mismatch(it) }?.let { return it }
            e = e.cause
        }
        return null
    }

    /**
     * Reads the certificate a server presents, without trusting it: the
     * handshake is always aborted, so nothing is sent. For trust on first
     * use, when someone types a LAN address and there's no pin yet.
     */
    fun capture(host: String, port: Int, timeoutMs: Int = 3000): String? {
        val tm = object : X509TrustManager {
            override fun checkServerTrusted(chain: Array<out X509Certificate>?, authType: String?) {
                val leaf = chain?.firstOrNull() ?: throw CertificateException("no certificate")
                throw Captured(fingerprint(leaf))
            }

            override fun checkClientTrusted(chain: Array<out X509Certificate>?, authType: String?) {
                throw CertificateException("not for clients")
            }

            override fun getAcceptedIssuers(): Array<X509Certificate> = emptyArray()
        }
        val ctx = SSLContext.getInstance("TLS").apply { init(null, arrayOf(tm), null) }
        val socket = java.net.Socket()
        return try {
            socket.connect(InetSocketAddress(host, port), timeoutMs)
            socket.soTimeout = timeoutMs
            val ssl = ctx.socketFactory.createSocket(socket, host, port, true) as SSLSocket
            ssl.use { it.startHandshake() }
            null  // unreachable: the trust manager always throws
        } catch (e: Exception) {
            var c: Throwable? = e
            var found: String? = null
            while (c != null && found == null) {
                if (c is Captured) found = c.fingerprint
                c = c.cause
            }
            found
        } finally {
            runCatching { socket.close() }
        }
    }

    /**
     * The X.509 certificate behind a WebView SSL error. API 29 has a getter;
     * before that the framework's saved state carries the DER bytes.
     */
    fun x509(cert: SslCertificate?): X509Certificate? {
        cert ?: return null
        if (Build.VERSION.SDK_INT >= 29) cert.x509Certificate?.let { return it }
        return fromSavedState(runCatching { SslCertificate.saveState(cert) }.getOrNull())
    }

    /** The certificate in [SslCertificate.saveState]'s bundle, which keeps the DER under "x509-certificate". */
    fun fromSavedState(saved: Bundle?): X509Certificate? {
        val der = saved?.getByteArray("x509-certificate") ?: return null
        return runCatching {
            CertificateFactory.getInstance("X.509").generateCertificate(ByteArrayInputStream(der)) as X509Certificate
        }.getOrNull()
    }

    /** Whether a WebView may proceed past [cert]: only if it's exactly the pinned certificate. */
    fun webViewMayProceed(cert: SslCertificate?, pin: String?): Boolean {
        if (!isFingerprint(pin)) return false
        val x = x509(cert) ?: return false
        return MessageDigest.isEqual(fingerprint(x).toByteArray(), pin!!.toByteArray())
    }

    private const val LAN_CONNECT_S = 3L
}
